codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -378,16 +384,34 @@ protected Http2FrameWriter frameWriter() {
      * Writes a {@code RST_STREAM} frame to the remote endpoint and updates the connection state appropriately.
      */
     @Override
-    public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
-            ChannelPromise promise) {
-        Http2Stream stream = connection().stream(streamId);
+    public ChannelFuture writeRstStream(final ChannelHandlerContext ctx, int streamId, long errorCode,
+            final ChannelPromise promise) {
+        final Http2Stream stream = connection().stream(streamId);
+        if (stream == null) {
+            // Don't write a RST_STREAM frame if we are not aware of the stream.
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
+                    "Stream %d not found while writing RST_STREAM frame", streamId));
+        }
+        if (stream.isResetSent()) {
+            // Don't write a RST_STREAM frame if we already wrote a RST_STREAM frame
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
```

Comment:
> Is this call considered in isolation, or should we be considering the state of the connection relative to the interface.

If we fail or succeed we always react based on the state of the connection.

> but do you have a specific use case (or call stack) that may enlighten me?

We have this in GRPC actually where close could be called multiple times on different ends and thus it would lead to several calls to `writeRstStream`.